dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/notification/ProgramStageNotificationMessageRenderer.java

```diff
@@ -173,4 +175,9 @@ private static String filterValue( EventDataValue dv )
 
         return value != null ? value : MISSING_VALUE_REPLACEMENT;
     }
+
+    private static DataElement getDataElement( String dataElementUid )
```

Comment:
Though we are caching dataElements, the fact that we are hitting - at least to do the cache - DB per dataElement is not efficient.

The notification service is triggered in the context of an event, where we know the programStage. The moment we know the programStage, we can cache its dataElements. Actually we don't really need to do this as it is just one event we are dealing with here.

Reply:
Fixed.